src/rt/cpp/cown.h

```diff
@@ -0,0 +1,165 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#include <functional>
+#include <tuple>
+#include <utility>
+#include <verona.h>
+
+class cown_ptr_trait
+{
+private:
+  cown_ptr_trait() {}
+
+  template<typename T>
+  friend class cown_ptr;
+};
+
+/**
+ * Smart pointer to represent shared access to a cown.
+ * Can only be used asychronously with `when` to get
+ * underlying access.
+ */
+template<typename T>
+class cown_ptr : cown_ptr_trait
+{
+private:
+  /**
+   * Internal Verona runtime cown for this type.
+   */
+  class ActualCown : public VCown<ActualCown>
+  {
+  private:
+    T value;
+
+  public:
+    ActualCown(T&& t) : value(std::forward<T>(t)) {}

```

Comment:
Is this supposed to be a forwarding constructor (in which case it should probably be a variadic template) or a move constructor that assumes a constructed `T`?  A comment would be good to see which.